src/include/rbtree.h

```diff
@@ -1,17 +1,173 @@
+ * If the range overlaps with existing nodes, it intelligently merges or modifies nodes to maintain non-overlapping, contiguous ranges.
+ * This includes extending, splitting, or merging nodes based on the overlap conditions and data consistency. 
+ * It handles multiple edge cases:
+ * 1. Overlap with same data: The existing node will be extended.
+ * 2. Overlap with different data: The existing node will split into two or three nodes depending on the position of the new address range, 
+ *    and the overlapped part will be overwritten with the data of the new address range.
+ * 
+ * (case 1)
+ *       +---------+---------+---------+---------+---------+
+ *       |                   data A                        |
+ *       +---------+---------+---------+---------+---------+
+ *                  ^.....       overlap      ....^
+ *                  +---------+---------+---------+
+ *                  |            data B           |
+ *                  +---------+---------+---------+
+ * 
+ *                                ||
+ *                               \||/
+ *                                \/
+ *   
+ *       +---------+ +---------+---------+---------+ +---------+
+ *       | data A  | |       data B                | |  data A |
+ *       +---------+ +---------+---------+---------+ +---------+
+ * 
+ * (case 2)
+ *       +---------+---------+---------+---------+---------+
+ *       |                   data A                        |
+ *       +---------+---------+---------+---------+---------+
+ *                           ^.....       overlap      ....^
+ *                           +---------+---------+---------+
+ *                           |            data B           |
+ *                           +---------+---------+---------+
+ * 
+ *                                ||
+ *                               \||/
+ *                                \/
+ *   
+ *       +---------+---------+ +---------+---------+---------+
+ *       |      data A       | |           data B            |
+ *       +---------+---------+ +---------+---------+---------+
+ * 
+ * (case 3)
+ *       +---------+---------+---------+---------+---------+
+ *       |                   data A                        |
+ *       +---------+---------+---------+---------+---------+
+ *                           ^.....       overlap      ....^
+ *                            +---------+---------+---------+---------+---------+
+ *                            |                   data B                        |
+ *                            +---------+---------+---------+---------+---------+
+ * 
+ *                                ||
+ *                               \||/
+ *                                \/
+ *   
+ *       +---------+---------+ +-----------------------------+---------+---------+
+ *       |      data A       | |                     data B                      |
+ *       +---------+---------+ +-------------------+---------+---------+---------+
```

Comment:
This is nitpicking as well: the `data B` block is missing some `+` signs (they got replaced by `-`).